### ISSUE TITLE
Proper tail calls are not supported properly

### ISSUE THREAD

**Issue by mnieper** [6to5/6to5#737]
6to5 fails to convert this:

```javascript
function f(n, m) {
  if (n === 0) {
    return m;
  }
  return g(n - 1, n * m); 
}

function g(a, b) {
  return f(a, b);
}
```

**Comment by sebmck** [comment]
See the prior discussion in #714, notably [this](https://github.com/6to5/6to5/pull/714#issuecomment-73434308) comment:

> the one included in this PR isn't in the 6to5 core because there are issues with nested tail calls.

**Comment by mnieper** [comment]
Thanks for the quick reply. You might want add a warning to http://6to5.org/docs/learn-es6/#tail-calls that 6to5 does not have proper tail calls (yet)

**Comment by sebmck** [comment]
Ah yeah, good call. There was initially an "unsupported" note (6to5/6to5.github.io@aefa5c0ecdf6a6a5221696ced53e1ed4e60b651d) but I remove it after #714, that pull was reverted and I hadn't gotten around to updating the docs. I'll do that now, thanks!

**Comment by disnet** [comment]
Should tail calls be disabled by default too? Updating to the latest version of babel broke my code because I had one of the unsupported tail calls.

**Comment by sebmck** [comment]
@disnet Only self recursing tail calls are optimised which isn't related to this issue. What code did it break?

**Comment by disnet** [comment]
This [function](https://github.com/mozilla/sweet.js/blob/es6-modules/src/expander.js#L638). Before updating to 4.0.1 it wasn't being rewritten and then after updating it was and got into stack overflow. I fixed it by adding "es6.tailCall" to the blacklist.

**Comment by sebmck** [comment]
What version were you on before? 4.x didn't introduce anything new, the only reason for the major bump was the name change.

**Comment by disnet** [comment]
I was on 2.12.5.

**Comment by sebmck** [comment]
Weird, going over the transpiled output there doesn't appear to be anything wrong. Do you have a stacktrace or steps to reproduce so I can track it down?

**Comment by disnet** [comment]
You can checkout commit [a785ee88d645f94b098c4b200a9d9cc54972b82b](https://github.com/mozilla/sweet.js/commit/a785ee88d645f94b098c4b200a9d9cc54972b82b) (right before I added tail calls to the blacklist) and then do a `npm test` to see the stack trace. I'll see if I can pair it down to a smaller test case.

**Comment by disnet** [comment]
We've got a recursive call inside another function [here](https://github.com/mozilla/sweet.js/blob/es6-modules/src/expander.js#L726). Would this cause problems?

**Comment by sebmck** [comment]
Nope, it wouldn't. sweet.js repo just finished cloning, I'll have a look.

**Comment by sebmck** [comment]
I cannot for the life of me debug this because the code's so dense, sorry!

**Comment by disnet** [comment]
Haha no worries, that's the heart of things and crazy complicated. I'll open another issue if I figure out what is going on.
